versioned_hdf5/wrappers.py

```diff
@@ -726,6 +749,18 @@ def __getitem__(self, args, new_dtype=None):
         # Return arr as a scalar if it is shape () (matching h5py)
         return arr[()]
 
+    @with_phil
+    def __getitem__(
+        self,
+        args: Union[slice, Slice, Tuple, tuple, h5r.RegionReference],
+        new_dtype: Optional[str] = None,
```

Comment:
Is anyone using `new_dtype`? Currently it does not seem to work:
```
In [1]: with TempDirCtx(DIR_cluster_tmp()) as d:
   ...:     with h5py.File(d/'foo.h5', 'w') as f:
   ...:         vf = VersionedHDF5File(f)
   ...:         with vf.stage_version('r0') as sv:
   ...:             sv.create_dataset('values', data=np.arange(10))
   ...:             values = sv['values']
   ...:             print(values[::2, np.dtype('double')])
   ...:
[PYFLYBY] from deshaw.constants.dirs import DIR_cluster_tmp
[PYFLYBY] from deshaw.file.filename import TempDirCtx
[PYFLYBY] from versioned_hdf5 import VersionedHDF5File
[PYFLYBY] import h5py
[PYFLYBY] import numpy as np
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
Cell In[1], line 7
      5 sv.create_dataset('values', data=np.arange(10))
      6 values = sv['values']
----> 7 print(values[::2, np.dtype('double')])

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:950, in InMemoryArrayDataset.__getitem__(self, item)
    949 def __getitem__(self, item):
--> 950     return self.array.__getitem__(item)

IndexError: only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices

In [2]: with TempDirCtx(DIR_cluster_tmp()) as d:
   ...:     with h5py.File(d/'foo.h5', 'w') as f:
   ...:         vf = VersionedHDF5File(f)
   ...:         with vf.stage_version('r0') as sv:
   ...:             sv.create_dataset('values', data=np.arange(10))
   ...:         with vf.stage_version('r1') as sv:
   ...:             values = sv['values']
   ...:             print(values[::2, np.dtype('double')])
   ...:
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[2], line 8
      6 with vf.stage_version('r1') as sv:
      7     values = sv['values']
----> 8     print(values[::2, np.dtype('double')])

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:1129, in DatasetWrapper.__getitem__(self, index)
   1128 def __getitem__(self, index):
-> 1129     return self.dataset.__getitem__(index)

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/versioned_hdf5/wrappers.py:658, in InMemoryDataset.__getitem__(self, args, new_dtype)
    655 idx = ndindex(args).expand(self.shape)
    657 if self.id.can_read_direct:
--> 658     return super().__getitem__(idx.raw)
    660 arr = np.ndarray(idx.newshape(self.shape), new_dtype, order='C')
    662 for chunk in self.chunks.as_subchunks(idx, self.shape):

File h5py/_objects.pyx:54, in h5py._objects.with_phil.wrapper()

File h5py/_objects.pyx:55, in h5py._objects.with_phil.wrapper()

File /usr/local/python/python-3.11/std/lib64/python3.11/site-packages/h5py/_hl/dataset.py:768, in Dataset.__getitem__(self, args, new_dtype)
    766 if self._fast_read_ok and (new_dtype is None):
    767     try:
--> 768         return self._fast_reader.read(args)
    769     except TypeError:
    770         pass  # Fall back to Python read pathway below

File h5py/_selector.pyx:361, in h5py._selector.Reader.read()

File h5py/_selector.pyx:107, in h5py._selector.Selector.apply_args()

ValueError: 2 indexing arguments for 1 dimensions
```
Probably we should just drop that?

Reply:
I'd be fine dropping this - this is actually the first time I've ever seen a `__getitem__` implementation that takes more than one argument anyway.